src/overload_hessian.jl

```diff
@@ -1,85 +1,193 @@
 ## 1-to-1
-for fn in ops_1_to_1
-    @eval function Base.$fn(t::T) where {T<:GlobalHessianTracer}
-        if is_seconder_zero_global($fn)
-            if is_firstder_zero_global($fn)
-                return empty(T)
-            else
-                return t
-            end
+function hessian_tracer_1_to_1(
+    t::T, is_firstder_zero::Bool, is_seconder_zero::Bool
+) where {T<:HessianTracer}
+    if is_seconder_zero
+        if is_firstder_zero
+            return empty(T)
         else
-            return T(t.grad, t.hess ∪ (t.grad × t.grad))
+            return t
         end
+    else
+        return T(gradient(t), hessian(t) ∪ (gradient(t) × gradient(t)))
     end
 end
 
-## 2-to-1
-for fn in ops_2_to_1
-    @eval function Base.$fn(a::T, b::T) where {G,H,T<:GlobalHessianTracer{G,H}}
-        grad = empty(G)
-        hess = empty(H)
-        if !is_firstder_arg1_zero_global($fn)
-            grad = union(grad, a.grad) # TODO: use union!
-            union!(hess, a.hess)
-        end
-        if !is_firstder_arg2_zero_global($fn)
-            grad = union(grad, b.grad) # TODO: use union!
-            union!(hess, b.hess)
-        end
-        if !is_seconder_arg1_zero_global($fn)
-            union!(hess, a.grad × a.grad)
-        end
-        if !is_seconder_arg2_zero_global($fn)
-            union!(hess, b.grad × b.grad)
-        end
-        if !is_crossder_zero_global($fn)
-            union!(hess, (a.grad × b.grad) ∪ (b.grad × a.grad))
-        end
-        return T(grad, hess)
+for fn in ops_1_to_1
+    @eval function Base.$fn(t::HessianTracer)
+        return hessian_tracer_1_to_1(
+            t, is_firstder_zero_global($fn), is_seconder_zero_global($fn)
+        )
+    end
+    @eval function Base.$fn(d::D) where {P,T<:HessianTracer,D<:Dual{P,T}}
+        x = primal(d)
+        p_out = Base.$fn(x)
+        t_out = hessian_tracer_1_to_1(
+            tracer(d), is_firstder_zero_local($fn, x), is_seconder_zero_local($fn, x)
+        )
+        return Dual(p_out, t_out)
     end
+end
 
-    @eval function Base.$fn(t::T, ::Number) where {G,H,T<:GlobalHessianTracer{G,H}}
-        if is_seconder_arg1_zero_global($fn)
-            if is_firstder_arg1_zero_global($fn)
-                return empty(T)
-            else
-                return t
-            end
-        else
-            return T(t.grad, t.hess ∪ (t.grad × t.grad))
-        end
+## 2-to-1
+function hessian_tracer_2_to_1(
+    a::T,
+    b::T,
+    is_firstder_arg1_zero::Bool,
+    is_seconder_arg1_zero::Bool,
+    is_firstder_arg2_zero::Bool,
+    is_seconder_arg2_zero::Bool,
+    is_crossder_zero::Bool,
+) where {G,H,T<:HessianTracer{G,H}}
+    grad = empty(G)
```

Comment:
here again, should we also trace the gradient with the utilities defined in `overloads_gradient.jl`?

Reply:
Done.

Reply:
`overloads_gradient.jl` only contains definitions on `GradientTracer` and I would have to put unions everywhere save on 2 lines of code duplication.

Reply:
Then let's duplicate. What's important is that we make sure the exact same operations are performed as in the gradient tracing case